File: Haskell/problem040.hs
{-
Problem 40 - Project Euler
http://projecteuler.net/index.php?section=problems&id=040
-}

concatIntegers nmax = concat $ map show [0..nmax]

problem040 = 
            let
                ci = concatIntegers 1000000
                d = [1, 10, 100, 1000, 10000, 100000, 1000000]
                ds = map ((!!) ci) d
            in
                product (map (read . return) ds :: [Int])

main = putStrLn $ "problem040: answer " ++ show problem040

File: Kotlin/problem040.kt
// Problem 40 - Project Euler
// http://projecteuler.net/index.php?section=problems&id=40
fun main(args: Array<String>) {
    var n = 1
    var str = ""
    var d = 1
    var ans = listOf<Char>()
    while (str.length < 100000) {
        str += n.toString()

        if (str.length > d) {
            ans = ans.plus(str[d-1])
            d *= 10
        }
        n++
    }
    println(ans)
    println(ans.map { it.toInt() - 48 }.fold(1) { acc, x -> acc * x })
}
